File: netscope-go/internal/engine/replay.go
package engine

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/drvoss/netscope/netscope-go/internal/model"
)

// Deterministic replay (spec §9).
//
// A scenario file supplies a virtual monotonic clock and a fixed stream of
// ProbeResults. Replaying it opens no sockets and reads no real clock, so both
// binaries can be fed the identical stream and their canonical JSON snapshots
// compared byte for byte. Mirroring directory names does not demonstrate parity;
// this does.
//
// The scenario schema is intentionally shallow so the C++ side can parse it with
// a small hand-rolled reader. Keep it in sync with
// netscope-cpp/src/core/replay.cpp.

type ScenarioTarget struct {
	Input        string `json:"input"`
	IP           string `json:"ip"`
	Family       string `json:"family"`
	ResolvedAtMs int64  `json:"resolvedAtMs"`
}

type ScenarioLocal struct {
	Interface    string   `json:"interface"`
	Address      string   `json:"address"`
	Gateway      string   `json:"gateway"`
	DNSServers   []string `json:"dnsServers"`
	DefaultRoute string   `json:"defaultRoute"`
	PublicIP     string   `json:"publicIp"`
	Note         string   `json:"note"`
}

type ScenarioHealth struct {
	HTTPStatus    int      `json:"httpStatus"`
	HTTPLatencyMs *float64 `json:"httpLatencyMs"`
	HTTPNote      string   `json:"httpNote"`
	TCPPort       int      `json:"tcpPort"`
	TCPOpen       bool     `json:"tcpOpen"`
	TCPNote       string   `json:"tcpNote"`
}

// ScenarioStep is a tagged union discriminated by Kind:
// "probe" | "enrich" | "trace-round" | "snapshot" | "event" | "pause" | "reprobe"
type ScenarioStep struct {
	Kind      string  `json:"kind"`
	TMs       int64   `json:"tMs"`
	TTL       int     `json:"ttl"`
	Attempt   uint64  `json:"attempt"`
	Outcome   string  `json:"outcome"`
	Responder string  `json:"responder"`
	RTTMs     float64 `json:"rttMs"`
	IP        string  `json:"ip"`
	RDNS      string  `json:"rdns"`
	ASN       string  `json:"asn"`
	Org       string  `json:"org"`
	EventKind string  `json:"eventKind"`
	Text      string  `json:"text"`
}

type Scenario struct {
	Name     string         `json:"name"`
	Mode     string         `json:"mode"`
	Target   ScenarioTarget `json:"target"`
	Local    ScenarioLocal  `json:"local"`
	Health   ScenarioHealth `json:"health"`
	Steps    []ScenarioStep `json:"steps"`
	EmitAtMs int64          `json:"emitAtMs"`
}

// LoadScenario reads a scenario file.
func LoadScenario(path string) (*Scenario, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var s Scenario
	if err := json.Unmarshal(b, &s); err != nil {
		return nil, fmt.Errorf("%s: %w", path, err)
	}
	if s.Mode == "" {
		s.Mode = string(model.ModeRaw)
	}
	return &s, nil
}

// Replay drives the engine through a scenario and returns the final snapshot.
func Replay(s *Scenario) *model.Snapshot {
	family := model.Family(s.Target.Family)
	if family == "" {
		family = model.FamilyIP4
	}
	e := New(model.Target{
		Input:      s.Target.Input,
		IP:         s.Target.IP,
		Family:     family,
		ResolvedAt: time.Duration(s.Target.ResolvedAtMs) * time.Millisecond,
	}, model.ProbeMode(s.Mode))

	e.SetLocal(model.LocalInfo{
		Interface:    s.Local.Interface,
		Address:      s.Local.Address,
		Gateway:      s.Local.Gateway,
		DNSServers:   s.Local.DNSServers,
		DefaultRoute: s.Local.DefaultRoute,
		PublicIP:     s.Local.PublicIP,
		Note:         s.Local.Note,
	})
	e.SetHealth(model.Health{
		HTTPStatus:  s.Health.HTTPStatus,
		HTTPLatency: s.Health.HTTPLatencyMs,
		HTTPNote:    s.Health.HTTPNote,
		TCPPort:     s.Health.TCPPort,
		TCPOpen:     s.Health.TCPOpen,
		TCPNote:     s.Health.TCPNote,
	})

	for _, st := range s.Steps {
		at := time.Duration(st.TMs) * time.Millisecond
		switch st.Kind {
		case "probe":
			out := model.Outcome(st.Outcome)
			r := model.ProbeResult{
				ID: model.ProbeID{
					Generation: e.Generation(),
					Family:     family,
					TTL:        st.TTL,
					Attempt:    st.Attempt,
				},
				Outcome:   out,
				Responder: st.Responder,
				SentAt:    at,
			}
			if out.Answered() {
				r.RTT = time.Duration(st.RTTMs * float64(time.Millisecond))
				r.RecvAt = at + r.RTT
			} else {
				r.RecvAt = at + time.Duration(e.Cadence().ProbeTimeoutMs)*time.Millisecond
			}
			r.Note = st.Text
			e.Ingest(r)
		case "enrich":
			e.ApplyEnrich(st.IP, st.RDNS, st.ASN, st.Org)
		case "trace-round":
			e.EndTraceRound(at)
		case "snapshot":
			// Classification is stateful (hysteresis), so the scenario controls
			// exactly when snapshots are taken.
			e.Snapshot(at)
		case "event":
			var ttl *int
			if st.TTL != 0 {
				ttl = model.Int(st.TTL)
			}
			e.AddEvent(at, model.EventKind(st.EventKind), ttl, st.Text)
		case "pause":
			e.TogglePause(at)
		case "reprobe":
			e.Reprobe(at)
		}
	}

	return e.Snapshot(time.Duration(s.EmitAtMs) * time.Millisecond)
}

File: netscope-cpp/src/replay.cpp
#include "replay.h"

#include <fstream>
#include <sstream>

#include "engine.h"
#include "json.h"

namespace netscope {
namespace {

bool readFile(const std::string& path, std::string& out, std::string& err) {
    std::ifstream in(path, std::ios::binary);
    if (!in) {
        err = "cannot open " + path;
        return false;
    }
    std::ostringstream os;
    os << in.rdbuf();
    out = os.str();
    return true;
}

}  // namespace

std::shared_ptr<const Snapshot> replayFile(const std::string& path, std::string& err) {
    std::string text;
    if (!readFile(path, text, err)) return nullptr;

    auto doc = parseJson(text, err);
    if (!doc) {
        err = path + ": " + err;
        return nullptr;
    }
    if (doc->type != JsonValue::Type::Object) {
        err = path + ": top level is not an object";
        return nullptr;
    }

    const JsonValue* targetObj = doc->find("target");
    Target target;
    Family family = Family::IP4;
    if (targetObj && targetObj->type == JsonValue::Type::Object) {
        target.input = targetObj->strOr("input");
        target.ip = targetObj->strOr("ip");
        family = familyFromName(targetObj->strOr("family", "ip4"));
        target.family = family;
        target.resolvedAt = ms(targetObj->intOr("resolvedAtMs", 0));
    }

    Engine eng(target, modeFromName(doc->strOr("mode", "raw")));

    if (const JsonValue* l = doc->find("local"); l && l->type == JsonValue::Type::Object) {
        LocalInfo info;
        info.interfaceName = l->strOr("interface");
        info.address = l->strOr("address");
        info.gateway = l->strOr("gateway");
        info.dnsServers = l->stringsAt("dnsServers");
        info.defaultRoute = l->strOr("defaultRoute");
        info.publicIp = l->strOr("publicIp");
        info.note = l->strOr("note");
        eng.setLocal(std::move(info));
    }

    if (const JsonValue* h = doc->find("health"); h && h->type == JsonValue::Type::Object) {
        Health health;
        health.httpStatus = static_cast<int>(h->intOr("httpStatus", 0));
        health.httpLatencyMs = h->optNum("httpLatencyMs");
        health.httpNote = h->strOr("httpNote");
        health.tcpPort = static_cast<int>(h->intOr("tcpPort", 0));
        health.tcpOpen = h->boolOr("tcpOpen", false);
        health.tcpNote = h->strOr("tcpNote");
        eng.setHealth(std::move(health));
    }

    const JsonValue* steps = doc->find("steps");
    if (steps && steps->type == JsonValue::Type::Array) {
        for (const JsonValue& st : steps->array) {
            if (st.type != JsonValue::Type::Object) continue;
            const std::string kind = st.strOr("kind");
            const Dur at = ms(st.intOr("tMs", 0));

            if (kind == "probe") {
                ProbeResult r;
                r.id.generation = eng.generation();
                r.id.family = family;
                r.id.ttl = static_cast<int>(st.intOr("ttl", 0));
                r.id.attempt = static_cast<std::uint64_t>(st.intOr("attempt", 0));
                r.outcome = outcomeFromName(st.strOr("outcome", "Timeout"));
                r.responder = st.strOr("responder");
                r.sentAt = at;
                if (answered(r.outcome)) {
                    r.rtt = fromMs(st.numOr("rttMs", 0));
                    r.recvAt = at + r.rtt;
                } else {
                    r.recvAt = at + ms(eng.cadence().probeTimeoutMs);
                }
                r.note = st.strOr("text");
                eng.ingest(r);

            } else if (kind == "enrich") {
                eng.applyEnrich(st.strOr("ip"), st.strOr("rdns"), st.strOr("asn"), st.strOr("org"));

            } else if (kind == "trace-round") {
                eng.endTraceRound(at);

            } else if (kind == "snapshot") {
                // Classification is stateful (hysteresis), so the scenario
                // controls exactly when snapshots are taken.
                eng.snapshot(at);

            } else if (kind == "event") {
                std::optional<int> ttl;
                const int t = static_cast<int>(st.intOr("ttl", 0));
                if (t != 0) ttl = t;
                eng.addEvent(at, eventKindFromName(st.strOr("eventKind", "start")), ttl,
                             st.strOr("text"));

            } else if (kind == "pause") {
                eng.togglePause(at);

            } else if (kind == "reprobe") {
                eng.reprobe(at);
            }
        }
    }

    return eng.snapshot(ms(doc->intOr("emitAtMs", 0)));
}

}  // namespace netscope
